x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/OrdinalsGroupingOperator.java

```diff
@@ -375,16 +377,27 @@ void addInput(IntVector docs, Page page) {
                     currentReader = new BlockOrdinalsReader(docValuesSupplier.get(), blockFactory);
                 }
                 try (IntBlock ordinals = currentReader.readOrdinalsAdded1(docs)) {
-                    for (int p = 0; p < ordinals.getPositionCount(); p++) {
-                        int start = ordinals.getFirstValueIndex(p);
-                        int end = start + ordinals.getValueCount(p);
-                        for (int i = start; i < end; i++) {
-                            long ord = ordinals.getInt(i);
+                    IntVector ordinalsVector = ordinals.asVector();
+                    if (ordinalsVector != null) {
```

Comment:
So that's how you do it with blocks

Reply:
Yeah. The generated code for `eval` functions does it for you.

There's a little difference though - vectors are always dense. So it won't hit as often as the single valued lucene optimization.